h5py/h5fd.pyx

```diff
@@ -77,3 +77,106 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
 # Flag for tracking allocation of space in file
 LOG_ALLOC     = H5FD_LOG_ALLOC      # 0x4000
 LOG_ALL       = H5FD_LOG_ALL        # (H5FD_LOG_ALLOC|H5FD_LOG_TIME_IO|H5FD_LOG_NUM_IO|H5FD_LOG_FLAVOR|H5FD_LOG_FILE_IO|H5FD_LOG_LOC_IO)
+
+
+# Implementation of 'fileobj' Virtual File Driver: HDF5 Virtual File
+# Layer wrapper over Python file-like object.
+# https://support.hdfgroup.org/HDF5/doc1.8/TechNotes/VFL.html
+
+# HDF5 events (read, write, flush, ...) are dispatched via
+# H5FD_class_t (struct of callback pointers, H5FD_fileobj_*). This is
+# registered as the handler for 'fileobj' driver via H5FDregister.
+
+# H5FD_fileobj_open callback acts first, expecting "filename" from
+# Python side in form
+#   hex(id(file)).encode('ASCII')
+# and retruning struct H5FD_fileobj_t (descendant of base H5FD_t)
+# which will hold file state. Other callbacks receive H5FD_fileobj_t
+# and operate on f.fileobj. If successful, callbacks must return zero;
+# otherwize non-zero value.
+
+
+# H5FD_t of file-like object
+ctypedef struct H5FD_fileobj_t:
+  H5FD_t base  # must be first
+  PyObject* fileobj
+  haddr_t eoa
+
+
+# A minimal subset of callbacks is implemented. Non-essential
+# parameters (dxpl, type) are ignored.
+
+from cpython cimport Py_INCREF, Py_DECREF
+from libc.stdint cimport *
+from libc.stdio cimport *
+
+
+cdef H5FD_fileobj_t *H5FD_fileobj_open(const char *name, unsigned flags, hid_t fapl, haddr_t maxaddr):
+    cdef unsigned long long int p
+    cdef H5FD_fileobj_t *f
+    if sscanf(name, "%llx", &p):
+        f = <H5FD_fileobj_t *>malloc(sizeof(H5FD_fileobj_t))
+        f.fileobj = <PyObject *>p
+        Py_INCREF(<object>f.fileobj)
+        f.eoa = 0
+        return f
+
+cdef herr_t H5FD_fileobj_close(H5FD_fileobj_t *f) except -1:
+    Py_DECREF(<object>f.fileobj)
+    free(f)
+    return 0
+
+cdef haddr_t H5FD_fileobj_get_eoa(const H5FD_fileobj_t *f, H5FD_mem_t type):
+    return f.eoa
+
+cdef herr_t H5FD_fileobj_set_eoa(H5FD_fileobj_t *f, H5FD_mem_t type, haddr_t addr) except -1:
+    f.eoa = addr
+    return 0
+
+cdef haddr_t H5FD_fileobj_get_eof(const H5FD_fileobj_t *f, H5FD_mem_t type) except -1:
```

Comment:
`haddr_t` is unsigned and `get_eof` cannot fail in the hdf5 API. To make sure it cannot fail, other drivers just cache the eof. Currently `-1` gets turned into a vary large but otherwise valid value for the eof.

Reply:
No, `get_eof` can actually return `HADDR_UNDEF` to signal failure:
https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5FD.c#1226

Reply:
Interesting, I haven't seen a driver that does this. I stand corrected. I looked into this because when `tell` fails this raises a system error. I wonder if we are not propagating this error through correctly in the Cython wrappers. Could we use the symbolic constant `HADDR_UNDEF` to make this more clear?

Reply:
Will do. As for Python error propagation, it seems that `H5Epush2` is the way to go (although it will add a lot of boilerplate compared to elegant Cython's `except`).

Reply:
The hdf5 codebase uses macros for this, maybe we can use them also.